src/realm/group_writer.cpp

```diff
@@ -376,31 +397,27 @@ ref_type GroupWriter::write_group()
     // clobering the previous database version. Note, however, that this risk
     // would only have been present in the non-transactional case where there is
     // no version tracking on the free-space chunks.
+#if REALM_ALLOC_DEBUG
+    std::cout << "    Freelist size after allocations: " << m_free_positions.size() << std::endl;
+#endif
+    // do not preserve order of freelist, just append new blocks. Order will be restored
+    // anyway by the merge_free_lists in the start of next commit.
+    for (const auto& free_space : m_not_free_in_file) {
+        m_free_positions.add(free_space.ref);
+        m_free_lengths.add(free_space.size);
+        if (is_shared)
+            m_free_versions.add(free_space.released_at_version);
+    }
     for (const auto& free_space : new_free_space) {
-        ref_type ref = free_space.ref;
-        size_t size = free_space.size;
-        // We always want to keep the list of free space in sorted order (by
-        // ascending position) to facilitate merge of adjacent segments. We
-        // can find the correct insert postion by binary search
-        size_t ndx = m_free_positions.lower_bound_int(ref);
-        if (ndx > 0) {
-            ref_type prev_ref = to_ref(m_free_positions.get(ndx - 1));
-            size_t prev_size = to_size_t(m_free_lengths.get(ndx - 1));
-            REALM_ASSERT_RELEASE_EX(prev_ref + prev_size <= ref, prev_ref, prev_size, ref, ndx, m_free_positions.size());
```

Comment:
These assertions should be replaced with something else

Reply:
in progress...